components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/interceptors/auth/BasicAuthInterceptor.java

```diff
@@ -0,0 +1,315 @@
+     * @param username     username
+     * @return true if user is authorized, false otherwise.
+     */
+    private boolean validateRoles(Message inMessage, UserRealm userRealm, String tenantDomain, String username) {
+        String basePath = (String) inMessage.get(Message.BASE_PATH);
+        String path = (String) inMessage.get(Message.PATH_INFO);
+        String verb = (String) inMessage.get(Message.HTTP_REQUEST_METHOD);
+        String resource = path.substring(basePath.length() - 1);
+        String[] userRoles;
+        Map<String, String> restAPIScopes;
+        //get all the URI templates of the REST API from the base path
+        Set<URITemplate> uriTemplates = getURITemplatesForBasePath(basePath);
+
+        //iterate through all the URITemplates to get the relevant URI template and get the scopes attached to validate
+        for (Object template : uriTemplates.toArray()) {
+            org.wso2.uri.template.URITemplate templateToValidate;
+            Map<String, String> var = new HashMap<>();
+            String templateString = ((URITemplate) template).getUriTemplate();
+            try {
+                templateToValidate = new org.wso2.uri.template.URITemplate(templateString);
+
+                //check if the current URITemplate matches with the resource and verb of the API request
+                if (templateToValidate.matches(resource, var) && verb != null
+                        && verb.equalsIgnoreCase(((URITemplate) template).getHTTPVerb())) {
+
+                    //get the scope list of the matched URITemplate
+                    List<Scope> resourceScopeList = ((URITemplate) template).retrieveAllScopes();
+
+                    //Continue the role check only if the invoking resource URI template has roles
+                    if (!resourceScopeList.isEmpty()) {
+                        //get the configured RESTAPIScopes map for the tenant from cache or registry
+                        restAPIScopes = getRESTAPIScopesForTenant(tenantDomain);
+                        if (restAPIScopes != null) {
+                            //get the current role list of the user from local user store manager 
+                            userRoles = userRealm.getUserStoreManager()
+                                    .getRoleListOfUser(MultitenantUtils.getTenantAwareUsername(username));
+                            if (userRoles != null) {
+                                return validateUserRolesWithRESTAPIScopes(resourceScopeList, restAPIScopes,
+                                        userRoles, username, path, verb);
+                            } else {
+                                log.error("Error while validating roles. Invalid user roles found for user: "
+                                        + username);
+                                return false;
+                            }
+                        } else {
+                            //Error while getting the RESTAPIScopes
```

Comment:
Shall we have an error log here

Reply:
The restAPIScopes will only be null if an exception is thrown from https://github.com/wso2/carbon-apimgt/pull/6763/files#diff-57f87d2d93009cdff072814012b16769R237. If so, the error log is printed from  https://github.com/wso2/carbon-apimgt/pull/6763/files#diff-57f87d2d93009cdff072814012b16769R244 .